odoo/tools/mail.py

```diff
@@ -533,6 +533,26 @@ def decode_smtp_header(smtp_header):
         return sep.join([ustr(x[0], x[1]) for x in text])
     return u''
 
+
+def decode_smtp_address_header(smtp_header):
+    """Behave like ``decode_smtp_header`` but safely reintroduce dbquotes
+    around Bob in 'Bob <bob@example.com>'. This ensure a name that
+    contains a coma is correctly decoded."""
+    if isinstance(smtp_header, Header):
+        smtp_header = ustr(smtp_header)
+    if smtp_header:
+        bparts = decode_header(smtp_header.replace('\r', ''))
+        uparts = []
+        for bpart, encoding in bparts:
+            upart = ustr(bpart, encoding)
+            if encoding not in (None, 'ascii') and upart[0] != '"' and upart[-1] != '"':
```

Comment:
The detection of 'ascii' encoding is incorrect, RFC2047 encoding could be done with ASCII (cfr the examples with US-ASCII at the end of RFC2047 for example).

I think the heuristics could be very simple: if the word was encoded (encoding is set), it should be re-quoted, no matter what.

Section 5.3 of RFC2047 explicitly forbids encoding the `addr-spec` part, so the address will always come in an unencoded part (``enc==None``).

Based on this, here's a proposition for a simpler implementation, as a slightly modified version of `decode_smtp_address_header()` to support a "quoted" mode.
You would simply pass `quoted=True` in the relevant parts of `message_parse()`.

```py
def decode_smtp_header(smtp_header, quoted=False):
    """Returns unicode() string conversion of the given encoded smtp header
    text. email.header decode_header method return a decoded string and its
    charset for each decoded part of the header. This method unicodes the
    decoded header and join them in a complete string.

    :param bool quoted: when True, encoded words in the header will be turned into RFC822
        quoted-strings after decoding, which is appropriate for address headers
    """
    if isinstance(smtp_header, Header):
        smtp_header = ustr(smtp_header)
    if smtp_header:
        pairs = decode_header(smtp_header.replace('\r', ''))
        tokens = []
        for token, enc in pairs:
            token = ustr(token, enc)
            if enc and quoted:
                # re-quote the encoded word to form an RFC822 quoted-string
                token = email_addr_escapes_re.sub(r'\\\g<0>', token)
                tokens.append('"%s"' % token)
            else:
                # plain word
                tokens.append(token)
        return ''.join(tokens)
    return ''
```

Reply:
Noteworthy: it is perfectly valid to have a space-separated sequence of encoded words, such as ``=?utf-8?b?RG/DqSwg?= =?US-ASCII?Q?Joe?= <joe@ex.com>``. In that case there are 2 choices: quote them all together after decoding them, or quote them individually and put the quoted-strings next to each other. Both cases are valid from an RFC point of view, and the second option is easier to implement, even if it gives a surprising result:

```py
# First option - quote after merging (slightly harder to implement)
In [1]: decode_smtp_header('=?utf-8?b?RG/DqSwg?= =?US-ASCII?Q?Joe?= <joe@ex.com>', True)
Out[1]: '"Doé, Joe" <joe@ex.com>'

# Second option - quote each word individually. Looks weird but it's valid.
In [1]: decode_smtp_header('=?utf-8?b?RG/DqSwg?= =?US-ASCII?Q?Joe?= <joe@ex.com>', True)
Out[1]: '"Doé, ""Joe" <joe@ex.com>'
```

Second result looks weird, but it parses and gets displayed exactly like the first result on RFC-compliant software. It's not a major concern anyway, as the most common case is to find a single encoded word for the whole "name" part.

Reply:
Do you have an idea how we could safely decode mixed ascii/unicode in the display name like `Joé Doe <joe@ex.com>` or `Joe Doé <joe@ex.com>` so it quotes the entire display name instead of *just* the RFC-2047 atom ?